### ISSUE TITLE
microk8s disable ha-cluster clears data

### ISSUE THREAD

**Issue by sstubbs** [ubuntu/microk8s#1577]
When disabling ha-cluster the database is cleared. I expected something along these lines to happen but just in case it's not expected behaviour I thought I would add it here.

**Comment by richardbeumer** [comment]
Hello,

I'm facing the same issue for a small home test lab. I have a 3 node cluster (single Hyper-V host with 3 Microk8s VM nodes). I want to replace the Hyper-V host with a single bare metal Intell NUC.

I was hoping to add the NUC in the cluster, disabling HA and then removing the other nodes. But after disabling HA on the NUC it is no longer a cluster member. Instead it is a single node cluster on its own.

Is there another way to migrate from a 4 node HA cluster to single node cluster?

**Comment by ktsakalozos** [comment]
@richardbeumer you do not need to disable HA. You can add your new node to the cluster and then remove one-by-one the other nodes. The new NUC node should take over the Kubernetes datastore. @freeekanayaka correct me if I am wrong.

**Comment by freeekanayaka** [comment]
@ktsakalozos that's correct, although I'm not sure this particular scenario was tested. It should work.

**Comment by richardbeumer** [comment]
@ktsakalozos  gave it a try. But when I remove the last node my cluster becomes complete unresponsive. I have to restore the last node for the cluster to become responsive again.

**Comment by balchua** [comment]
I tried it too.  There is a check on `leave` command preventing the user to remove a node when it is the only leader remaining.

**Comment by knittl** [comment]
I just ran into the same "problem". I was investigating CPU issues and thought: I run microk8s on a single node, I can disable the ha-cluster plugin, maybe it is using all the CPU, it can't hurt – well, wrong. It DOES hurt.

I think this is **very critical**. There is no warning whatsoever :confused: All deployments, services, all pods, and all persistent volumes are _gone_ after disabling the plugin. Re-enabling the plugin does not bring back anything, the node remains empty :(

There should be a big fat warning or at least having the user confirm that they are about to delete all their data.

**Comment by giannoug** [comment]
Exact same issue for me. MicroK8s has 100% usage on a poor Intel Atom CPU, read that disabling ha-cluster helps, disabled ha-cluster, all data gone. No backups since this is a homelab machine 🗡️ (didn't hurt that much other than I don't have time to rebuilt it now)

**Comment by giosdas** [comment]
Hello,
I'm facing the same issue. Is it possible to add, at least, a warning about it when disabling ha-cluster?
Or require an argument like "--force" ?

Thanks
Regards

**Comment by giosdas** [comment]
Hello,

When I try to remove the HA from the third and last node, I received an error message about that it is the node with the last copy of k8s data and it prevent me from disabling ha.

So I revert all by re-enabling the HA cluster on the first and second nodes and readding them to the third node.

In the end the config is never lost, is kept on the last node.
Also in the cluster I use rook-ceph and no volume data has been lost.

Regards

**Comment by ktsakalozos** [comment]
We have PR [1] ready that adds a `--force` requirement to the `disable ha-cluster` when you have a single node.

That should be released with v1.21.

[1] https://github.com/ubuntu/microk8s/pull/2143

**Comment by Azbesciak** [comment]
Hello, is there any way to set up default services/pods/... after disabling HA? After it is hardly usable.
So far I have dumped the whole default deployment but it requires some cleaning which I want to avoid.

**Comment by Azbesciak** [comment]
Is the default config available somewhere to apply? Or as I did, I need to dump it?

**Comment by ktsakalozos** [comment]
I am sorry, what is the default config?

**Comment by Azbesciak** [comment]
The One after fresh installation, but without HA

**Comment by ktsakalozos** [comment]
There is no config to apply after the ha is disabled. The cluster should be completely empty. You should only see the kubernetes service, nothing else.

**Comment by Azbesciak** [comment]
In my opinion, there should be some guidance at least how to configure cluster without HA, despite it known performance issues. I exported the node configuration before disabling HA, with `kubectl get all --all-namespaces -o yaml > ...`, disabled HA, applied the file (after 5 mins of cleanup) and... service accounts not found, because this resource is not available in `all` macro. On the new node, I exported them and applied again base configuration with these... so far so good.
Now lets add new node with `add-node`. On the worker node (fresh installation) I paste the command from master (`microk8s join ...` and guess what?
> Connection failed. Not possible to join. This is not an HA MicroK8s cluster (501).

Why is it so hard to set up without HA? is it a feature that also requires at least 3 nodes, which I suppose some clients do not have, or rather a must-have?

**Comment by busbyjon** [comment]
Data is also removed if you have an old node and install "ha-cluster"

No warnings, nothing. All config is removed.  Very frustrating!
